File: 1 Month Preparation Kit/Week 3/Simple Text Editor/Simple Text Editor.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <stack>
using namespace std;

void appendString(string& s) {
    string str;
    getline(cin >> ws,str);
    s.append(str);
}

void deleteLastKChar(string& s) {
    int k;
    cin >> k;
    s.erase(s.size()-k,k);     
}

void printKChar(string& s) {
    int k;
    cin >> k;
    cout << s[k-1] << endl;
}

void undo(string& s,stack<string>& undoStack) {
    if(undoStack.size() > 1 ) {
        undoStack.pop();
        s = undoStack.top();
    } else {
        s = "";
    }
}

void editString(string& s,char t,stack<string>& undoStack) {

    switch (t) {
    case '1':
        appendString(s); 
        undoStack.push(s);
        break;
    case '2':
        deleteLastKChar(s);
        undoStack.push(s);
        break;
    case '3':
        printKChar(s);
        break;
    case '4':
        undo(s, undoStack);
        break;
    default:
        cout << "ERR 404" << endl;
        break;
    }
}


int main() {
    /* Enter your code here. Read input from STDIN. Print output to STDOUT */
    stack<string> undoStack;
    int q;
    cin >> q;
    string current;
    undoStack.push(current);
    for(int i = 0; i < q;i++) {
        char t;
        cin >> t;
        editString(current, t,undoStack);   
    }
    return 0;
}

File: 1 Month Preparation Kit/Week 3/Simple Text Editor/Simple Text Editor.py
# Enter your code here. Read input from STDIN. Print output to STDOUT

if __name__ == '__main__':
    q = int(input())
    
    stack = []
    state = ''
    
    for i in range(q):
        query = input().split()
        if query[0] == '1':
            stack.append(state)
            state += query[1]
        elif query[0] == '2':
            stack.append(state)
            state = state[:len(state) - int(query[1])]
        elif query[0] == '3':
            print(state[int(query[1]) - 1])
        elif query[0] == '4':
            state = stack.pop()
